fix various macOS specific problems

ara4n opened the pull request:
surely the maximum size for the text should be the width of the parent
(i.e. topLayout) minus the width of the avatar layout, which is max_height_?

mujx commented on src/RoomInfoListItem.cc:
```diff
@@ -77,12 +77,12 @@ RoomInfoListItem::RoomInfoListItem(RoomInfo info, QWidget *parent)
 
 	roomName_ = new QLabel(info_.name(), textWidget_);
-	roomName_->setMaximumSize(parent->width() - 10, 20);
+	roomName_->setMaximumSize(parent->width() - max_height_, 20);
```
I don't think those are necessary, it just squashes the labels further inside. The 
```
setMaximumSize(parent->width(), max_height_);
``` 
line, was probably the bugfix. Could you confirm on this?

ara4n replied:
I reverted the `roomName`, `roomTopic` and `setElidedText` sizing changes, which caused everything to break again:

<img width="1178" alt="screen shot 2017-04-30 at 12 42 46" src="https://cloud.githubusercontent.com/assets/1294269/25564096/e9b2b108-2da2-11e7-9172-89989c607c51.png">

I then re-added the setElidedText change, just to check if it was just the eliding operation which was mis-sized, which improved things a bit but still broke with rooms with a long room name (Megolm test, in this example):

<img width="1178" alt="screen shot 2017-04-30 at 12 44 37" src="https://cloud.githubusercontent.com/assets/1294269/25564100/0c9df0a6-2da3-11e7-8ebf-e49be5cc512c.png">

...so I'm pretty sure that the patch here is correct.  My rationale is that `parent->width()` is returning the width of the whole `RoomInfoListItem`, so you should be sizing the width of the `roomName` and `roomTopic` boxes to that width, minus the width of the avatar (which is set to `max_height_`).

Then again, I know nothing about Qt :D

mujx replied:
Yes you're right. But on Linux the situation is a little different. The room topic is cut off a lot sooner.

![room-list](https://cloud.githubusercontent.com/assets/6430350/25564645/6f613cd0-2dc0-11e7-8259-c28ad45c9982.png)


The problem with the room name is that the font size isn't taken into consideration. Same goes for the room topic. We should use `QFontMetrics` to cut off the text. Maybe the screen resolution has an impact on this and this was the problem all along.

![test](https://cloud.githubusercontent.com/assets/6430350/25564633/26f259de-2dc0-11e7-821b-a8c277f330e0.png)

mujx commented on src/ImageOverlayDialog.cc:
```diff
@@ -38,6 +38,23 @@ ImageOverlayDialog::ImageOverlayDialog(QPixmap image, QWidget *parent)
+	// It might be slightly cleaner to do:
+	//
+	// #ifdef Q_WS_MACX
+	// #include <Carbon/Carbon.h>
```
Will this work by default on OSX? If yes we can uncomment this section. Btw how other platforms would be affected by this?

ara4n replied:
The problem on macOS here is that for whatever reason, when the ImageOverlayDialog is dismissed, the fullscreen state isn't fully reset for the app - meaning you lose the application menu bar for the rest of the app's lifetime.

I assume this is either a bug in Qt where it fails to call the MacOS API `ShowMenuBar()` to reset the UI when closing a QWindow with WindowState `QWindow::FullScreen`, or there's some missing housekeeping that lets Qt reset the app correctly.

I think the second option is more likely, so I'm faking the cleanup by forcing the QDialog to de-fullscreen itself before it gets destroyed, which correctly resets the app's menu bar.  This could also affect other platforms where Qt implements some application-wide behaviour for fullscreen (e.g. hiding the menu bar), so I think this is the better solution than doing the macOS-specific hack of breaking the Qt abstraction and directly calling the Carbon API needed to hack around the problem.

By the way, why does the constructor `raise()`?

mujx replied:
The solution is good enough for me. Since I'm pretty new to Qt myself, I might have missing something or the full screen widget isn't implemented properly. So maybe it's not a Qt bug but we have missed the proper API calls.

You can remove `raise()`, it's probably a leftover from previous iterations.

mujx commented:
I'm a little confused with the last three commits. Probably something went wrong with the rebase because `raise()` is not removed.

The `parent->width() - max_height_` is correct. I just added a `-5` on my branch so the text doesn't touch the right border.  The only thing missing from the PR is to replace the stylesheet font size as shown in `fix/roomlist-text`. You can also revert the `setMaximumSize` as it keeps the widget smaller (there are examples of this on the previous comments).

jansol commented:
> I just added a -5 on my branch so the text doesn't touch the right border.

Please make that a const variable (`PADDING_RIGHT` or so)

ara4n commented:
I added raise() back in when merging because you hadn't removed it when adding the close hook, and because i realised i had assumed raise() was a python-style "raise a Qt exception" rather than the "raise this window to the top"... which seems like a reasonable thing to do.  Happy to kill it again though if that's the right thing to do :)

I reverted to allow for a clean merge with your branch.

i'll test to see if reverting the setMaximumSize is ok on OSX - but be aware that your fix-room-list branch does have nasty 4px of horizontal overscroll when i tested it earlier.

-- 
Matthew Hodgson
Matrix.org

> On 30 Apr 2017, at 20:59, mujx <notifications@github.com> wrote:
> 
> I'm a little confused with the last three commits. Probably something went wrong with the rebase because raise() is not removed.
> 
> The parent->width() - max_height_ is correct. I just added a -5 on my branch so the text doesn't touch the right border. The only thing missing from the PR is to replace the stylesheet font size as shown in fix/roomlist-text. You can also revert the setMaximumSize as it keeps the widget smaller (there are examples of this on the previous comments).
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

mujx commented:
Thanks for clearing that up. I thought you were going to integrate the font part in your branch since most of the work has already been done there. Could you post a screenshot with the 4px problem?

When you are ready just squash the commits.

ara4n commented:
(out-of-band it seems the 4px problem is indeed the need for setMaximumSize on the RoomInfoListItem)

Slightly dubious about squashing the commits (why not just do a squash merge?), but I guess i can do so on the assumption that nobody else is playing with this branch...

ara4n commented:
okay, i've tried to rebase and squash the null commits, although it looks to have made a bit of a mess :( hopefully it will merge simply though.

mujx commented:
Could you squash the commits into one? Btw the issue with the overlay image appearing tiny also happens on KDE/Plasma.

mujx commented on README.md:
````diff
@@ -42,6 +42,9 @@ $ brew update
 $ brew install qt5
 ```
 
+N.B. you will need to pass `-DCMAKE_PREFIX_PATH=/usr/local/Cellar/qt5/5.8.0_2`
````
Did you try the path on travis.yml? `CMAKE_PREFIX_PATH=/usr/local/opt/qt5`

ara4n commented:
ok, this is a complete mess now - opened #15 to replace it.
